Sema: Type check #_hasSymbol()

tshortli opened the pull request:
Implement type checking for the `#_hasSymbol` directive. The directive takes an expression which must resolve at compile time to a single concrete declaration. There are a few kinds of expressions which we intend to accept:

- Unapplied functions: `if #_hasSymbol(foo(_:))`
- Member references: `if #_hasSymbol(a.x)`
- `.self` expressions on static metatypes: `if #_hasSymbol(SomeEnum.self)`
- Dot syntax call expressions with unapplied functions: `if #_hasSymbol(SomeType.init(_:)`

Additionally, we diagnose when the concretely referenced declaration is not weak linked since this likely indicates that the programmer misidentified the declaration they wish to check or the build is not configured in a way such that the check will be meaningful.

Resolves rdar://99826340

tshortli committed:
Implement type checking for the `#_hasSymbol` directive. The directive takes an expression which must resolve at compile time to a single concrete declaration. There are a few kinds of expressions which we intend to accept:

- Unapplied functions: `if #_hasSymbol(foo(_:))`
- Member references: `if #_hasSymbol(a.x)`
- `.self` expressions on static metatypes: `if #_hasSymbol(SomeEnum.self)`
- Dot syntax call expressions with unapplied functions: `if #_hasSymbol(SomeType.init(_:)`

Additionally, we diagnose when the concretely referenced declaration is not weak linked since this likely indicates that the programmer misidentified the declaration they wish to check or the build is not configured in a way such that the check will be meaningful.

Resolves rdar://99826340

tshortli commented:
@swift-ci please test

nkcsgexi reviewed (approved):
Nice! The implementation is much simpler than I expected👍.

xedin commented:
Can this be used in a multi-statement closure?

tshortli commented:
> Can this be used in a multi-statement closure?

I think we'd ideally allow that, though I don't know if it's strictly necessary for now. Do you see something that needs to be different to support that?

xedin commented:
> Do you see something that needs to be different to support that?

Yes, statements in closure context are handled by the solver via conjunctions. You’d need to implement both constraint generation and solution application logic - see CSClosure.cpp visitIfStmt for how regular conditions are handled. Another question whether it should be supported by a result builder transform, that lives in BuilderTransform.cpp

tshortli commented:
Got it, thanks Pavel. For this PR I'll make sure we at least reject the syntax in those contexts so that it's clear they aren't supported yet but I'll also look into what it will take to support it in the future.

xedin commented:
Sounds good, thanks!

tshortli committed:
Sema: Temporarily reject if #_hasSymbol in a closure context.

We'll need to implement appropriate constraint generation to support them in such a context.

tshortli commented:
@swift-ci please smoke test

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -4365,10 +4365,16 @@ bool ConstraintSystem::generateConstraints(StmtCondition condition,
+    case StmtConditionElement::CK_HasSymbol: {
+      ASTContext &ctx = getASTContext();
+      ctx.Diags.diagnose(condElement.getStartLoc(),
+                         diag::has_symbol_unsupported_in_closures);
```
just FYI - since it's just temporary that's okay but if it wasn't it would have to be a fix instead of a diagnostic in here.

tshortli commented:
@swift-ci please smoke test

xedin commented:
@tshortli Testing of https://github.com/apple/swift/pull/60419 shows that use of `#_hasSymbol` in result builder body with `-enable-experimental-feature ResultBuilderASTTransform` results in `failed to produce a diagnostic` error, could you please take a look?

tshortli commented:
> @tshortli Testing of #60419 shows that use of `#_hasSymbol` in result builder body with `-enable-experimental-feature ResultBuilderASTTransform` results in `failed to produce a diagnostic` error, could you please take a look?

@xedin Will do. If you need to get unblocked feel free to comment out that part of the test for now.
